### PR TITLE
ProductManager now batches content fetching while updating products

### PR DESCRIPTION
**Pull request by Ceiu**
- ProductManager.applyProductUpdates now fetches product content
  in batches, and no longer repeatedly iterates existing product
  content mapping while applying content changes
- Added commitProduct and commitContent to ProductManager and ContentManager
  respectively
- Refactored refreshPoolsWithRegeneration to minimize the number of times
  we hit the database for a given operation
- Reverted several changes in ProductManager and ContentManager
- Added importProduct and importContent to ProductManager and
  ContentManager respectively
- Changed the signatures of several methods according with
  optimizations made throughout the codebase
- AbstractHibernateCurator.saveAll, .updateAll, and .saveOrUpdateAll
  no longer use a universal session.clear, and instead evict the
  persisted elements
- Updated entitler to batch product and content lookups while doing
  dev pool resolution
- Updated tests accordingly

### CODE REVIEW

**Comment by Ceiu** [comment]
Reviewers: Pay special attention to some import/refresh database effects. During testing, we noticed that product content mappings were persisted without ever updating the created or updated fields. Granted, we don't use them anymore anyway, it signals that there may be some underlying bugs left to resolve.

**Comment by Ceiu** [comment]
retest this please

**Comment by barnabycourt** [comment]
@Ceiu There are an awful lot of TODOs sprinkled around the code

File: server/src/main/java/org/candlepin/model/AbstractHibernateCurator.java [hunk 1]
```diff
@@ -503,56 +505,132 @@ public EntityManager getEntityManager() {
...
-                    i++;
                 }
-                if (flush) {
-                    session.flush();
-                    session.clear();
+            }
+            catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+            // catch (OptimisticLockException e) {
```
Comment (barnabycourt) [inline server/src/main/java/org/candlepin/model/AbstractHibernateCurator.java]: Please remove commented out block if it is no longer needed

File: server/src/main/java/org/candlepin/sync/EntitlementImporter.java [hunk 2]
```diff
@@ -124,17 +123,20 @@ public Subscription importObject(ObjectMapper mapper, Reader reader, Owner owner
...
-            entcnt++;
+            ++entcnt;
+
             CertificateSerial cs = new CertificateSerial();
             cs.setCollected(cert.getSerial().isCollected());
             cs.setExpiration(cert.getSerial().getExpiration());
             cs.setUpdated(cert.getSerial().getUpdated());
             cs.setCreated(cert.getSerial().getCreated());
-            csCurator.create(cs);
+            // csCurator.create(cs);
```
Comment (barnabycourt) [inline server/src/main/java/org/candlepin/sync/EntitlementImporter.java]: Does this need to be created here or are we relying on the transaction flush?

File: server/src/main/java/org/candlepin/sync/SubscriptionReconciler.java [hunk 3]
```diff
@@ -115,11 +126,11 @@ public void reconcile(Owner owner, List<Subscription> subsToImport, PoolCurator
...
             Pool local = null;
-            Map<String, Pool> map = existingPoolsByUpstreamPool.get(
-                subscription.getUpstreamPoolId());
+            Map<String, Pool> map = existingPoolsByUpstreamPool.get(subscription.getUpstreamPoolId());
             if (map == null) {
                 map = new HashMap<String, Pool>();
             }
+
             for (Pool localSub : map.values()) {
                 // TODO quantity
```
Comment (barnabycourt) [inline server/src/main/java/org/candlepin/sync/SubscriptionReconciler.java]: What remains TODO here?

Comment (Ceiu) [inline server/src/main/java/org/candlepin/sync/SubscriptionReconciler.java]: No idea. That's a preexisting comment.

**Comment by barnabycourt** [comment]
ACK from me, waiting on secondary review from @mstead

File: server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java [hunk 4]
```diff
@@ -183,60 +183,186 @@ public CandlepinPoolManager(
...
-                deletedSubs.add(subId);
+        log.debug("Importing {} product(s)...", productMap.size());
+        Map<String, Product> importedProducts = this.productManager
+            .importProducts(owner, productMap, importedContent);
+
+        for (String pid : existingProductIds) {
+            // FIXME: This is technically wrong, at the moment. Products which did not change will
+            // still be reported as changed if they existed in the DB prior to import.
+            changedProducts.add(importedProducts.get(pid));
+        }
```
Comment (mstead) [inline server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java]: After discussion, an RFE will be created to address this as a separate PR as this is minor.

**Comment by mstead** [comment]
@barnabycourt ACK. Merge away.

File: server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java (hunk 4 above)
Comment (barnabycourt) [inline server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java]: Trello card created to track: https://trello.com/c/8IEsuj5w
